File: backend/app/domain/enums.py
"""Enumerations from spec section 6.3.

No framework imports anywhere in `domain/` (AGENTS.md hard rule 4).
"""

from enum import StrEnum


class TournamentCategory(StrEnum):
    GRAND_SLAM = "GRAND_SLAM"
    ATP = "ATP"
    WTA = "WTA"


class Tour(StrEnum):
    ATP = "ATP"
    WTA = "WTA"


class Surface(StrEnum):
    HARD = "HARD"
    CLAY = "CLAY"
    GRASS = "GRASS"
    INDOOR_HARD = "INDOOR_HARD"


class RoundReached(StrEnum):
    """How far a player advanced. `WITHDREW` sits outside the order."""

    R128 = "R128"
    R64 = "R64"
    R32 = "R32"
    R16 = "R16"
    QF = "QF"
    SF = "SF"
    F = "F"
    CHAMPION = "CHAMPION"
    WITHDREW = "WITHDREW"


#: Earliest exit first. A later index means a deeper run.
ADVANCEMENT_ORDER: tuple[RoundReached, ...] = (
    RoundReached.R128,
    RoundReached.R64,
    RoundReached.R32,
    RoundReached.R16,
    RoundReached.QF,
    RoundReached.SF,
    RoundReached.F,
    RoundReached.CHAMPION,
)


def round_rank(round_reached: RoundReached | None) -> int:
    """Position in the advancement order. A withdrawal never advanced."""
    if round_reached is None or round_reached is RoundReached.WITHDREW:
        return -1
    return ADVANCEMENT_ORDER.index(round_reached)


class TournamentStatus(StrEnum):
    DRAFT = "DRAFT"
    PUBLISHED = "PUBLISHED"
    RUNNING = "RUNNING"
    FINISHED = "FINISHED"


class TournamentVisibility(StrEnum):
    PUBLIC = "PUBLIC"
    PRIVATE = "PRIVATE"


class BetGroupKind(StrEnum):
    TOURNAMENT = "TOURNAMENT"
    ROUND = "ROUND"


class BetGroupStatus(StrEnum):
    DRAFT = "DRAFT"
    OPEN = "OPEN"
    LOCKED = "LOCKED"
    SETTLED = "SETTLED"


class QuestionFamily(StrEnum):
    TYPED = "TYPED"
    GENERIC = "GENERIC"


class TypedQuestionKind(StrEnum):
    QF_PICKS = "QF_PICKS"
    SF_PICKS = "SF_PICKS"
    FINALIST_PICKS = "FINALIST_PICKS"
    CHAMPION = "CHAMPION"
    UNDERPERFORMER = "UNDERPERFORMER"
    BREAKOUT = "BREAKOUT"


class AnswerType(StrEnum):
    PLAYER = "PLAYER"
    MATCH_RESULT = "MATCH_RESULT"
    INTEGER = "INTEGER"
    CHOICE = "CHOICE"


class ParticipationStatus(StrEnum):
    ACTIVE = "ACTIVE"
    WITHDRAWN = "WITHDRAWN"


class GameListState(StrEnum):
    OPEN_FOR_SIGNUP = "OPEN_FOR_SIGNUP"
    RUNNING = "RUNNING"
    FINISHED = "FINISHED"


#: Legal set scores per match format (spec 5.3).
SET_SCORES_BEST_OF_THREE: tuple[str, ...] = ("2-0", "2-1")
SET_SCORES_BEST_OF_FIVE: tuple[str, ...] = ("3-0", "3-1", "3-2")


def legal_set_scores(best_of: int) -> tuple[str, ...]:
    """The results a match of this format can actually end in."""
    return SET_SCORES_BEST_OF_THREE if best_of == 3 else SET_SCORES_BEST_OF_FIVE

File: frontend/src/services/types/enums.ts
/**
 * Enumerations from spec section 6.3.
 *
 * Each is a frozen array plus a union type derived from it: the array drives
 * select options and fixture generation, the union drives the type system.
 */

export const TOURNAMENT_CATEGORIES = ['GRAND_SLAM', 'ATP', 'WTA'] as const;
export type TournamentCategory = (typeof TOURNAMENT_CATEGORIES)[number];

export const TOURS = ['ATP', 'WTA'] as const;
export type Tour = (typeof TOURS)[number];

export const SURFACES = ['HARD', 'CLAY', 'GRASS', 'INDOOR_HARD'] as const;
export type Surface = (typeof SURFACES)[number];

/**
 * How far a player got. Ordered from earliest exit to the title, so a later
 * index means a deeper run. `WITHDREW` sits outside the order and is listed
 * last by convention.
 */
export const ROUNDS_REACHED = ['R128', 'R64', 'R32', 'R16', 'QF', 'SF', 'F', 'CHAMPION', 'WITHDREW'] as const;
export type RoundReached = (typeof ROUNDS_REACHED)[number];

export const TOURNAMENT_STATUSES = ['DRAFT', 'PUBLISHED', 'RUNNING', 'FINISHED'] as const;
export type TournamentStatus = (typeof TOURNAMENT_STATUSES)[number];

export const TOURNAMENT_VISIBILITIES = ['PUBLIC', 'PRIVATE'] as const;
export type TournamentVisibility = (typeof TOURNAMENT_VISIBILITIES)[number];

export const BET_GROUP_KINDS = ['TOURNAMENT', 'ROUND'] as const;
export type BetGroupKind = (typeof BET_GROUP_KINDS)[number];

export const BET_GROUP_STATUSES = ['DRAFT', 'OPEN', 'LOCKED', 'SETTLED'] as const;
export type BetGroupStatus = (typeof BET_GROUP_STATUSES)[number];

export const QUESTION_FAMILIES = ['TYPED', 'GENERIC'] as const;
export type QuestionFamily = (typeof QUESTION_FAMILIES)[number];

/** Narrows the typed case. Null for a generic question (spec 6.2). */
export const TYPED_QUESTION_KINDS = [
  'QF_PICKS',
  'SF_PICKS',
  'FINALIST_PICKS',
  'CHAMPION',
  'UNDERPERFORMER',
  'BREAKOUT',
] as const;
export type TypedQuestionKind = (typeof TYPED_QUESTION_KINDS)[number];

export const ANSWER_TYPES = ['PLAYER', 'MATCH_RESULT', 'INTEGER', 'CHOICE'] as const;
export type AnswerType = (typeof ANSWER_TYPES)[number];

export const PARTICIPATION_STATUSES = ['ACTIVE', 'WITHDRAWN'] as const;
export type ParticipationStatus = (typeof PARTICIPATION_STATUSES)[number];

/** Best-of-three and best-of-five, as the number of sets needed to win. */
export const MATCH_FORMATS = [3, 5] as const;
export type MatchFormat = (typeof MATCH_FORMATS)[number];

/**
 * Legal set scores per match format (spec 5.3). A set score outside the list
 * for its draw is rejected by the API and blocked in the UI.
 */
export const SET_SCORES_BEST_OF_THREE = ['2-0', '2-1'] as const;
export const SET_SCORES_BEST_OF_FIVE = ['3-0', '3-1', '3-2'] as const;

export type BestOfThreeSetScore = (typeof SET_SCORES_BEST_OF_THREE)[number];
export type BestOfFiveSetScore = (typeof SET_SCORES_BEST_OF_FIVE)[number];
export type SetScore = BestOfThreeSetScore | BestOfFiveSetScore;

export function legalSetScores(format: MatchFormat): readonly SetScore[] {
  return format === 3 ? SET_SCORES_BEST_OF_THREE : SET_SCORES_BEST_OF_FIVE;
}
